File: src/2021/day03/python/solution.py
from functools import reduce
import math
def format_data(input: str) -> list[tuple[int, ...]]:
    return [
        tuple(map(lambda b: int(b), list(line)))
        for line in input.strip().splitlines()
    ]

def part1(input):
    numbers = format_data(input)
    mid = len(numbers) // 2
    results_sum_tuple = tuple(sum(elements) for elements in zip(*numbers))
    most_bit = tuple(map(lambda v: "1" if v >= mid else "0", results_sum_tuple))
    least_bit = tuple(map(lambda v: "1" if v < mid else "0", results_sum_tuple))
    most_int = int("".join(most_bit), 2 )
    least_int = int("".join(least_bit), 2 )

    return most_int * least_int


def get_most_num(input: list[tuple[int,...]]) -> int:
    potentials = input
    current_index = 0


    while len(potentials) > 1:
        mid = math.ceil(len(potentials) / 2)
        index_sum = tuple(sum(elements) for elements in zip(*potentials))[current_index]
        find_value = 1 if index_sum >= mid else 0
        potentials = [
            potential for potential in potentials
            if potential[current_index] == find_value
        ]
        current_index += 1

    return int("".join(map(str, potentials[0])), 2)

def get_least_num(input: list[tuple[int,...]]) -> int:
    potentials = input
    current_index = 0


    while len(potentials) > 1:
        mid = math.ceil(len(potentials) / 2)
        index_sum = tuple(sum(elements) for elements in zip(*potentials))[current_index]
        find_value = 1 if index_sum < mid else 0
        potentials = [
            potential for potential in potentials
            if potential[current_index] == find_value
        ]
        current_index += 1

    num = int("".join(map(str, potentials[0])), 2)
    return num


def part2(input):
    numbers = format_data(input)
    oxygen_rating = get_most_num(numbers)
    co2_scrubber = get_least_num(numbers)
    print(oxygen_rating, co2_scrubber)
    return oxygen_rating * co2_scrubber

File: src/2021/day03/typescript/solution.ts
/* eslint-disable @typescript-eslint/no-loop-func */
function formatData(input: string): number[][] {
  return input.split('\n').map((line) => line.split('').map(Number));
}

export function part1(input: string): number | string | undefined {
  const ratings = formatData(input);

  const totals = ratings[0].map((v, i) => {
    return ratings.map((rating) => rating[i]).reduce((acc, num) => acc + num);
  });

  const mid = ratings.length / 2;

  let most = [];
  let min = [];

  for (const total of totals) {
    if (total > mid) {
      most.push(1);
      min.push(0);
    } else {
      most.push(0);
      min.push(1);
    }
  }
  const mostValue = parseInt(most.join(''), 2);
  const minValue = parseInt(min.join(''), 2);
  return mostValue * minValue;
}

export function part2(input: string): number | string | undefined {
  const ratings = formatData(input);

  let potentials = [...ratings];
  let currentIndex = 0;

  let min = [],
    max = [];

  while (potentials.length > 1) {
    const currentTotal = potentials
      .map((potential) => potential[currentIndex])
      .reduce((sum, num) => num + sum);

    const currentMid = potentials.length / 2;

    if (currentTotal >= currentMid) {
      max.push(1);
    } else {
      max.push(0);
    }

    potentials = potentials.filter(
      (potential) => potential[currentIndex] === max.at(-1),
    );
    currentIndex++;
  }

  max = potentials[0];

  potentials = [...ratings];
  currentIndex = 0;
  while (potentials.length > 1) {
    const currentTotal = potentials
      .map((potential) => potential[currentIndex])
      .reduce((sum, num) => num + sum);

    const currentMid = potentials.length / 2;

    if (currentTotal >= currentMid) {
      min.push(0);
    } else {
      min.push(1);
    }

    potentials = potentials.filter(
      (potential) => potential[currentIndex] === min.at(-1),
    );

    currentIndex++;
  }

  min = potentials[0];

  const maxValue = parseInt(max.join(''), 2);
  const minValue = parseInt(min.join(''), 2);
  return maxValue * minValue;
}
